### PR TITLE
[ARM] az deployment group/sub what-if: Add --exclude-change-types/-x parameter

### PR DESCRIPTION
**Pull request by shenglol**
**Description**  
- Added `--exclude-change-types/-x` parameter to the `az deployment group/sub what-if` commands that allows users to exclude certain resource change types from What-If results, which addresses https://github.com/Azure/arm-template-whatif/issues/53.
- Added `--what-if-exclude-change-types/-x` parameter to the `az deployment group/sub create` commands.
- Also improved error reporting for the `az deployment * validate` commands.

**Testing Guide**  
- `az deployment group/sub what-if -f <path_to_deployment_file> -x`
- `az deployment * create -f <path_to_deployment_file> -c -x`

**History Notes**  

[ARM] az deployment group/sub what-if: Add --exclude-change-types/-x parameter.
[ARM] az deployment group/sub/mg/tenant create: Add --what-if-exclude-change-types/-x parameter.
[ARM] az deployment group/sub/mg/tenant validate: Show error messages in a better format.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S171

**Comment by shenglol** [comment]
Hi @zhoxing-ms , @qianwens and @Juliehzl can you help review the PR? Thanks in advance.

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 1]
```diff
@@ -285,6 +285,8 @@
     short-summary: The deployment name.
   - name: --what-if-result-format -r
     short-summary: The format of What-If results. Applicable when --confirm-with-what-if is set.
+  - name: --what-if-exclude-change-types -x
+    short-summary: Space-separated list of resource change types to be excluded from What-If results. Possible values are Create, Delete, Modify, Deploy, NoChange, and Ignore. Applicable when --confirm-with-what-if is set.
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py:289]: This is repeated with the help declared in `deployment_what_if_exclude_change_types_type`, and the help declared in `_params.py` will override the help in `_help.py`

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 2]
```diff
@@ -411,6 +413,8 @@
...
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py:417]: Same as above

File: src/azure-cli/azure/cli/command_modules/resource/_help.py [hunk 3]
```diff
@@ -567,6 +571,8 @@
     short-summary: The deployment mode.
   - name: --what-if-result-format -r
     short-summary: The format of What-If results. Applicable when --confirm-with-what-if is set.
+  - name: --what-if-exclude-change-types -x
+    short-summary: Space-separated list of resource change types to be excluded from What-If results. Possible values are Create, Delete, Modify, Deploy, NoChange, and Ignore. Applicable when --confirm-with-what-if is set.
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/_help.py:575]: Same as above

**Comment by zhoxing-ms** [comment]
@shenglol Could you add some tests for this change?

**Comment by shenglol** [comment]
> @shenglol Could you add some tests for this change?

Yes. Added unit tests for the new parameter.

File: src/azure-cli/azure/cli/command_modules/resource/_help.py (hunk 1 above)
Comment (shenglol) [reply src/azure-cli/azure/cli/command_modules/resource/_help.py:289]: Not sure if I misunderstood, but I think it's the other way it's the other way around, which is I what want. The short-summaries in `_help.py` override the help message specified in `deployment_what_if_exclude_change_types_type` - they end with `Applicable when --confirm-with-what-if is set.`.

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/resource/_help.py:289]: @shenglol I'm sorry, I typed it wrong. Doc: [help-layers](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_help.md#help-layers)
Actually, what I wanted to say was that there was no need to write help in two places, right?

Comment (shenglol) [reply src/azure-cli/azure/cli/command_modules/resource/_help.py]: No worries at all. Good point. I moved the help messages to `_params.py`.

**Comment by shenglol** [comment]
@zhoxing-ms can you take another look? Please let me know if there's any additional changes I need to make.

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 4]
```diff
@@ -64,6 +64,9 @@ def load_arguments(self, _):
     deployment_what_if_confirmation_type = CLIArgumentType(options_list=['--confirm-with-what-if', '-c'], action='store_true',
                                                            help='Instruct the command to run deployment What-If before executing the deployment. It then prompts you to acknowledge resource changes before it continues.',
                                                            is_preview=True, min_api='2019-07-01')
+    deployment_what_if_exclude_change_types_type = CLIArgumentType(nargs="+", options_list=['--exclude-change-types', '-x'],
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:67]: Can we use get_enum_type() for possible values?

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 5]
```diff
@@ -245,6 +248,9 @@ def load_arguments(self, _):
         c.argument('confirm_with_what_if', arg_type=deployment_what_if_confirmation_type)
         c.argument('what_if_result_format', options_list=['--what-if-result-format', '-r'],
                    arg_type=deployment_what_if_result_format_type)
+        c.argument('what_if_exclude_change_types', options_list=['--what-if-exclude-change-types', '-x'],
+                   arg_type=deployment_what_if_exclude_change_types_type,
+                   help="Space-separated list of resource change types to be excluded from What-If results. Possible values are Create, Delete, Modify, Deploy, NoChange, and Ignore. Applicable when --confirm-with-what-if is set.")
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:253]: The same

File: src/azure-cli/azure/cli/command_modules/resource/_params.py (hunk 4 above)
Comment (shenglol) [reply src/azure-cli/azure/cli/command_modules/resource/_params.py:67]: Thank you @Juliehzl . I changed it to use `get_enum_type()`. Your comments are always helpful. I was not aware of that `get_enum_types()` also works with nargs (I should have tried it).

File: src/azure-cli/azure/cli/command_modules/resource/_params.py (hunk 5 above)
Comment (shenglol) [reply src/azure-cli/azure/cli/command_modules/resource/_params.py:253]: Done.

**Pull request merged**